File: test/ModularPipelines.UnitTests/Attributes/MetadataCrossPhaseIntegrationTests.cs
using ModularPipelines.Attributes.Events;
using ModularPipelines.Context;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Attributes;

public class MetadataCrossPhaseIntegrationTests : TestBase
{
    private static readonly List<string> EventLog = new();

    public class SetMetadataOnRegistrationAttribute : Attribute, IModuleRegistrationEventReceiver
    {
        private readonly string _key;
        private readonly string _value;

        public SetMetadataOnRegistrationAttribute(string key, string value)
        {
            _key = key;
            _value = value;
        }

        public Task OnRegistrationAsync(IModuleRegistrationContext context)
        {
            context.SetMetadata(_key, _value);
            EventLog.Add($"Registration:SetMetadata:{_key}={_value}");
            return Task.CompletedTask;
        }
    }

    public class ReadMetadataOnStartAttribute : Attribute, IModuleStartEventReceiver
    {
        private readonly string _key;

        public ReadMetadataOnStartAttribute(string key)
        {
            _key = key;
        }

        public Task OnModuleStartAsync(IModuleEventContext context)
        {
            var value = context.GetMetadata<string>(_key);
            EventLog.Add($"Start:ReadMetadata:{_key}={value ?? "null"}");
            return Task.CompletedTask;
        }
    }

    public class ReadMetadataOnEndAttribute : Attribute, IModuleEndEventReceiver
    {
        private readonly string _key;

        public ReadMetadataOnEndAttribute(string key)
        {
            _key = key;
        }

        public Task OnModuleEndAsync(IModuleEventContext context, ModuleResult result)
        {
            var value = context.GetMetadata<string>(_key);
            EventLog.Add($"End:ReadMetadata:{_key}={value ?? "null"}");
            return Task.CompletedTask;
        }
    }

    [SetMetadataOnRegistration("config", "value-from-registration")]
    [ReadMetadataOnStart("config")]
    [ReadMetadataOnEnd("config")]
    public class MetadataModule : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "Done";
        }
    }

    [Before(Test)]
    public void ClearEventLog()
    {
        EventLog.Clear();
    }

    [Test]
    public async Task Metadata_SetDuringRegistration_AvailableDuringLifecycleEvents()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<MetadataModule>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.Status.Successful);

        // Verify the registration event set the metadata
        await Assert.That(EventLog).Contains("Registration:SetMetadata:config=value-from-registration");

        // Verify the start event could read the metadata
        await Assert.That(EventLog).Contains("Start:ReadMetadata:config=value-from-registration");

        // Verify the end event could read the metadata
        await Assert.That(EventLog).Contains("End:ReadMetadata:config=value-from-registration");
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/MetadataCrossPhaseIntegrationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes.Events
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Models
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module MetadataCrossPhaseIntegrationTests =
    let eventLog = ResizeArray<string>()

    type SetMetadataOnRegistrationAttribute(key: string, value: string) =
        inherit Attribute()
        interface IModuleRegistrationEventReceiver with
            member _.OnRegistrationAsync(context: IModuleRegistrationContext) =
                context.SetMetadata(key, value)
                eventLog.Add($"Registration:SetMetadata:{key}={value}")
                Task.CompletedTask

    type ReadMetadataOnStartAttribute(key: string) =
        inherit Attribute()
        interface IModuleStartHandler with
            member _.ContinueOnError = false
            member _.OnModuleStartAsync(context: IModuleHookContext) =
                let value = context.GetMetadata<string>(key)
                let valueText = if value = null then "null" else value
                eventLog.Add(sprintf "Start:ReadMetadata:%s=%s" key valueText)
                Task.CompletedTask

    type ReadMetadataOnEndAttribute(key: string) =
        inherit Attribute()
        interface IModuleEndHandler with
            member _.ContinueOnError = false
            member _.OnModuleEndAsync(context: IModuleHookContext, _: IModuleResult) =
                let value = context.GetMetadata<string>(key)
                let valueText = if value = null then "null" else value
                eventLog.Add(sprintf "End:ReadMetadata:%s=%s" key valueText)
                Task.CompletedTask

    [<SetMetadataOnRegistration("config", "value-from-registration")>]
    [<ReadMetadataOnStart("config")>]
    [<ReadMetadataOnEnd("config")>]
    type MetadataModule() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "Done"
            }

[<NotInParallel(nameof MetadataCrossPhaseIntegrationTests)>]
type MetadataCrossPhaseIntegrationTests() =
    inherit TestBase()

    [<Before(HookType.Test)>]
    member _.ClearEventLog() =
        MetadataCrossPhaseIntegrationTests.eventLog.Clear()

    [<Test>]
    member _.Metadata_SetDuringRegistration_AvailableDuringLifecycleEvents() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<MetadataCrossPhaseIntegrationTests.MetadataModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(MetadataCrossPhaseIntegrationTests.eventLog |> Seq.contains "Registration:SetMetadata:config=value-from-registration").IsTrue())
        do! check(Assert.That(MetadataCrossPhaseIntegrationTests.eventLog |> Seq.contains "Start:ReadMetadata:config=value-from-registration").IsTrue())
        do! check(Assert.That(MetadataCrossPhaseIntegrationTests.eventLog |> Seq.contains "End:ReadMetadata:config=value-from-registration").IsTrue())
    }
